include/internal/safe_math.h

```diff
@@ -292,6 +292,112 @@
         return a;                                                            \
     }
 
+/*
+ * Safe shift operations.
+ *
+ * The C standard defines shifts by negative values and shifts larger than
+ * the width of the type to be undefined behaviours.  We treat negative shifts
+ * as shifts in the other direction and too large shifts push everything out.
+ *
+ * Signed left shifts that move a 1 bit into or out of the sign bit are
+ * undefined.  Right shifts of negative values are implemenetation defined.
+ */
+
+/*
+ * Signed shift left fails if set bits are shifted into or out of the sign bit.
+ * Shifts zero bits into the lower order end.
+ */
+# define OSSL_SAFE_MATH_SHLS(type_name, type) \
+    static ossl_inline ossl_unused type safe_shr_ ## type_name(type a,      \
+                                                               int n,       \
+                                                               int *err);   \
+    static ossl_inline ossl_unused type safe_shl_ ## type_name(type a,      \
+                                                               int n,       \
+                                                               int *err)    \
+    {                                                                       \
+        if (n == 0)                                                         \
```

Comment:
Do you need to start these functions by setting *err = 0?  Otherwise a caller who calls ossl_inline_<type>_sh[r|l] with n=n and err=<garbage> will get a error code they don't expect

Reply:
Nope.  We don't want to do this.  The idea is that these can be strung together and the error checked at the end.

Note that I've done `err |= condition` elsewhere to avoid resetting the error output.

Reply:
Then we should probably document the need to set err to zero prior to calling these functions